Added json config for block temperatures

SteelGiant87 opened the pull request:
Here is an initial pull request for the configurable block temperatures addition. This should close #109 and close #123

I am happy to make structural changes as dictated by the TaN team, and obviously any fixes needed.

If the general form of this feature is acceptable, I can potentially expose similar configuration options for other aspects of the mod.

- Blocks and Material.FIRE can now have their temperatures controlled using json configs.
- Blocks can have a basic temperature specified, and also have specific temperatures defined depending on the block state.
- Configs are autogenerated if not already present.
- The default configuration performs the same function as the previous hardcoded temperature specification.

The block temperatures specified in the json are stored in a HashMap, which should hopefully make temperature lookups from block names fast even if there are a large number of user specified blocks.

SteelGiant87 committed:
Added json config for block temperatures

Blocks and Material.FIRE can now have their temperatures controlled
using json configs.
Blocks can have a basic temperature specified, and also have specific
temperatures defined depending on the block state.
Configs are autogenerated if not already present.
The default configuration performs the same function as the previous
hardcoded temperature specification.

Adubbz commented on src/main/java/toughasnails/config/GameplayConfigurationHandler.java:
```diff
@@ -8,26 +8,131 @@
+            materialTemperatureData = new MaterialTemperatureData();
+            
+            String materialConfigName = FilenameUtils.getFullPath(configFile.getPath()) + "materialTemperatureConfig.json";
+            
```
Would prefer it to be material_temperatures.json

SteelGiant87 replied:
ok

Adubbz commented on src/main/java/toughasnails/config/GameplayConfigurationHandler.java:
```diff
@@ -8,26 +8,131 @@
+            //Block temperature config:
+            blockTemperatureData = new HashMap<String, BlockTemperatureData>();
+            
+            String blockConfigName = FilenameUtils.getFullPath(configFile.getPath()) + "blockTemperatureConfig.json";
```
Would prefer it to be block_temperatures.json

SteelGiant87 replied:
ok

Adubbz commented on src/main/java/toughasnails/config/GameplayConfigurationHandler.java:
```diff
@@ -8,26 +8,131 @@
         {
             config = new Configuration(configFile);
             loadConfiguration();
+            
```
All of the json stuff should really go in a seperate class file. GameplayConfigurationHandler is intended specifically for gameplay.cfg

SteelGiant87 replied:
Ok, will move to new class.

Adubbz commented on src/main/java/toughasnails/temperature/MaterialTemperatureData.java:
```diff
@@ -0,0 +1,7 @@
+public class MaterialTemperatureData {
+
+	public float FIRE = 1.0F;
+	
```
This seems incomplete to me... Shouldn't you have a field for temperature and for the material itself? Also in terms of naming, we generally don't use caps unless they are either static final or it is an enum. This is *usually* consistent across most styles I've seen.

SteelGiant87 replied:
The only material that I was planning to do this for was Materials.FIRE, as it doesn't really make sense for the other materials except Materials.LAVA, and that can be achieved by specifying temperatures for Blocks.LAVA and Blocks.FLOWING_LAVA (with the advantage that source blocks and flowing blocks can be treated differently if desired).

I couldn't find a way of mapping from a string to a material name cleanly (Material.toString() isn't intelligible and I believe would change if properties of a material were to change), otherwise I would have allowed it to be specified for all materials anyway, even if it might not be wanted.

I will change it to lowercase.

Adubbz commented on src/main/java/toughasnails/config/GameplayConfigurationHandler.java:
```diff
@@ -8,26 +8,131 @@
+			}
         }
     }
 
```
Not just in this class, but other classes too, there's a lot of mixing and matching spaces and tabs. With TAN we only use spaces for all of our classes. There's an option in eclipse and Intelli J for this. This messes up rendering on Github in some browsers/operating systems.

SteelGiant87 replied:
Will change to spaces everywhere.

Adubbz commented on src/main/java/toughasnails/temperature/BlockTemperatureData.java:
```diff
@@ -0,0 +1,14 @@
+public class BlockTemperatureData {
+	public String blockName;
+	public float blockBaseTemperature;
+	public BlockTemperatureStateData[] stateTemperatures;
```
It seems you've recreated the properties of block states rather than just working natively with actual IBlockStates. Here's what we do in BoP for example: https://github.com/Glitchfiend/BiomesOPlenty/blob/BOP-1.11.x-6.0.x/src/main/java/biomesoplenty/common/util/config/BOPConfig.java#L447-L512

SteelGiant87 replied:
Ok, I will refactor it to store native blockstates.

I will use a copy of your BOP config to make a TAN config and address this comment and moving the json out of GamplayConfigurationHandler.

SteelGiant87 replied:
I'm in the process of changing to use IBlockStates. It now occurs to me that using an explicit blockstate would require the user to specify a lot of redundant data.

e.g. The toughasnails:campfire, which has properties "age" and "burning" wants to have temperature 12 when burning and 0 when not burning regardless of age.

To achieve this when specifying IBlockStates it would be necessary to specify temperatures as

toughasnails:campfire[age=0,burning=true] = 12
toughasnails:campfire[age=1,burning=true] = 12
toughasnails:campfire[age=2,burning=true] = 12
...
toughasnails:campfire[age=7,burning=true] = 12

Am I missing something? Is there a way to compare an IBlockState with just the "burning" property with a full blockstate of a campfire?

The first system I implemented allows specification of a temperature on a single property. This is admittedly less powerful, but I believe that the need to provide a full enumeration of all equivalent states is more painful and error prone than adding a way to specify AND on a set of properties in the first system. Alternatively, specify a full IBlockState but then separately specify which properties are to be used for comparison?

I do agree that whatever is implemented the data should be validated to check that it actually specifies extant blocks and states. This will require moving this from mod preInit to mod init so that all mods have actually had a chance to create their blocks and states.

Adubbz commented on src/main/java/toughasnails/temperature/modifier/ObjectProximityModifier.java:
```diff
@@ -82,35 +83,43 @@ public Temperature modifyTarget(World world, EntityPlayer player, Temperature te
         return new Temperature(newTemperatureLevel);
     }
-
+    
```
Unnecessary added space

SteelGiant87 replied:
I'll fix that

SteelGiant87 committed:
Merge remote-tracking branch 'refs/remotes/Glitchfiend/TAN-1.11.x-2.0.0' into TAN-1.11.x-2.0.0

SteelGiant87 committed:
Changes as specified in pull request

Changed to using native IBlockStates
Moved the config to a new class, TANConfig
Config is now validated against block states to check values specified
are valid.
Moved initialisation of TANConfig from preInit to init, as the config
now needs to have had all blocks for all mods created in preInit so that
it can validate the config against them.
The user now specifies an IBlockState to match against, and the set of
properties to consider for matching. If the properties list is empty
then only the block needs to match.
Changed case of FIRE to fire to be more consistent with project styling
Corrected indentation character from tabs to spaces to be more
consistent with project styling

SteelGiant87 committed:
no longer require use_properties

The config no longer has a field "use_properties" to specify which
properties of the state should be used to determine temperature.
Instead, the config can have properties in the state either set to a
wildcard ("*") or omitted. In both cases the effect is the same: any
value of that property is acceptable.
